File: robot_perception/robot_perception/lidar2image_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan, CameraInfo, Image
from cv_bridge import CvBridge
from yolov8_msgs.msg import DetectionArray
from message_filters import Subscriber, ApproximateTimeSynchronizer
import cv2
from tf2_ros import Buffer, TransformListener, TransformException
import numpy as np

from geometry_msgs.msg import PoseArray, Pose
import tf_transformations
import transforms3d


class Lidar2ImageNode(Node):

    def __init__(self):
        super().__init__('lidar2image_node')

        self.cv_bridge_ = CvBridge()

        # Camera intrinsics
        self.camera_info_ = None

        # TF buffer and listener
        self.tf_buffer_ = Buffer()
        self.tf_listener_ = TransformListener(self.tf_buffer_, self)

        # Subscribers using message_filters
        self.image_sub_ = Subscriber(self, Image, "bgr_image")
        self.detections_sub_ = Subscriber(self, DetectionArray, "/yolo/detections")
        self.scan_sub_ = Subscriber(self, LaserScan, "/scan")

        # Synchronizer
        self.detection_image_scan_sync = ApproximateTimeSynchronizer(
            [self.detections_sub_, self.image_sub_, self.scan_sub_], queue_size=100, slop=0.1)
        self.detection_image_scan_sync.registerCallback(self.detection_image_scan_callback)

        # Camera info subscriber
        self.caminfo_sub_ = self.create_subscription(
            CameraInfo, 'camera_info', self.caminfoCallback, 10)

        # Publisher for object positions
        self.object_positions_pub_ = self.create_publisher(PoseArray, 'detected_object_positions', 10)

        # Publisher for overlay image
        self.overlay_image_pub_ = self.create_publisher(Image, 'overlay_image', 10)

        # Publisher for labeled laser scan
        self.labeled_scan_pub_ = self.create_publisher(LaserScan, 'labeled_scan', 10)
        self.get_logger().info("Initialized lidar2image_node")

    def detection_image_scan_callback(self, detections_msg, image_msg, scan_msg):
        """
        Callback for synchronized detections, RGB images, and laser scans.
        """
        self.get_logger().info("Processing synchronized messages")

        # Ensure camera_info_ is available
        if self.camera_info_ is None:
            self.get_logger().warn("Camera info not yet received. Cannot process data.")
            return
        # self.get_logger().warn(f'intensities : {scan_msg.intensities}')

        # Get camera intrinsics
        fx = self.camera_info_['fx']
        fy = self.camera_info_['fy']
        cx = self.camera_info_['cx']
        cy = self.camera_info_['cy']

        # Get laser scan data and compute LiDAR points
        points_lidar, indices = self.compute_lidar_points(scan_msg)

        # Transform LiDAR points to camera frame
        time_stamp = rclpy.time.Time.from_msg(scan_msg.header.stamp)
        points_camera = self.transform_lidar_points_to_camera_frame(
            points_lidar, scan_msg.header.frame_id, image_msg.header.frame_id, time_stamp)

        if points_camera is None:
            # Transformation failed
            self.get_logger().warn("Could not transform lidar points to camera")
            return

        # Project points onto image plane
        u, v, x_cam, y_cam, z_cam, indices = self.project_points_to_image_plane(
            points_camera, fx, fy, cx, cy, image_msg, indices)

        # Initialize labels array
        labels = np.zeros_like(indices, dtype=int)

        # Associate projected points with detections and estimate object positions
        pose_array_msg, labels = self.associate_points_with_detections(
            u, v, x_cam, y_cam, z_cam, indices, labels, detections_msg, image_msg)
            

        # Publish the object positions
        self.object_positions_pub_.publish(pose_array_msg)

        # Convert image_msg to CV image
        cv_image = self.cv_bridge_.imgmsg_to_cv2(image_msg, desired_encoding='bgr8')

        # Draw all projected LiDAR points (before labeling)
        # for i in range(len(u)):
        #     cv2.circle(cv_image, (u[i], v[i]), 2, (100, 100, 100), -1)  # Blue points

        # Draw bounding boxes and labels for red and green objects
        for detection in detections_msg.detections:
            class_id = detection.class_id

            # Process only red and green objects (assuming class IDs 0 and 1)
            if class_id in [1, 2]:
                bbox = detection.bbox
                image_width = image_msg.width
                image_height = image_msg.height

                # Bounding box coordinates
                xmin = int(bbox.center.position.x - bbox.size.x / 2)
                ymin = int(bbox.center.position.y - bbox.size.y / 2)
                xmax = int(bbox.center.position.x + bbox.size.x / 2)
                ymax = int(bbox.center.position.y + bbox.size.y / 2)

                # Clip coordinates to image dimensions
                xmin = max(0, xmin)
                ymin = max(0, ymin)
                xmax = min(image_width - 1, xmax)
                ymax = min(image_height - 1, ymax)

                # Choose color based on class ID
                # if class_id == 0:
                #     color = (0, 0, 0)  # Black
                #     label_text = "Black"
                if class_id == 1:
                    color = (0, 255, 0)  # Green
                    label_text = "Green"
                elif class_id == 2:
                    color = (0, 0, 255)  # Red
                    label_text = "Red"
                else:
                    continue

                cv2.rectangle(cv_image, (xmin, ymin), (xmax, ymax), color, 2)
                cv2.putText(cv_image, label_text, (xmin, ymin - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, color, 2)

        # Draw labeled LiDAR points for red and green objects
        for i in range(len(u)):
            # if labels[i] != 0:
            # if labels[i] == 0:
            #     color = (0, 0, 0)  # Black
            if labels[i] == 1:
                color = (0, 255, 0)  # Green
            elif labels[i] == 2:
                    color = (0, 0, 255)  # red
            else:
                # color = (255, 0, 0)  # Blue
                continue
            
            cv2.circle(cv_image, (u[i], v[i]), 2, color, -1)
        # self.get_logger().info(
        #                         f"Intenisty: {labels}"
        #                     )
        # Save and publish the image
        overlay_image_msg = self.cv_bridge_.cv2_to_imgmsg(cv_image, encoding='bgr8')
        overlay_image_msg.header = image_msg.header
        self.overlay_image_pub_.publish(overlay_image_msg)

        # Create labels_for_scan array
        labels_for_scan = np.zeros(len(scan_msg.ranges), dtype=float)

        # Map labels back to the original scan indices
        labels_for_scan[indices] = labels

        # Create new LaserScan message
        labeled_scan = LaserScan()
        labeled_scan.header = scan_msg.header
        labeled_scan.angle_min = scan_msg.angle_min
        labeled_scan.angle_max = scan_msg.angle_max
        labeled_scan.angle_increment = scan_msg.angle_increment
        labeled_scan.time_increment = scan_msg.time_increment
        labeled_scan.scan_time = scan_msg.scan_time
        labeled_scan.range_min = scan_msg.range_min
        labeled_scan.range_max = scan_msg.range_max
        labeled_scan.ranges = scan_msg.ranges
        labeled_scan.intensities = labels_for_scan.tolist()  # Convert to list
        # self.get_logger().info(
        #                         f"Intenisty: {labeled_scan.intensities}"
        #                     )
        # Publish the labeled scan
        self.labeled_scan_pub_.publish(labeled_scan)

    def compute_lidar_points(self, scan_msg):
        """
        Compute LiDAR points from LaserScan message.
        """
        # Get laser scan data
        ranges = np.array(scan_msg.ranges)
        angles = np.arange(len(ranges)) * scan_msg.angle_increment + scan_msg.angle_min
        indices = np.arange(len(ranges))

        # Filter out invalid ranges
        valid_indices = np.isfinite(ranges) & (ranges >= scan_msg.range_min) & (ranges <= scan_msg.range_max)
        ranges = ranges[valid_indices]
        angles = angles[valid_indices]
        indices = indices[valid_indices]

        # Compute LiDAR points in LiDAR frame
        x_lidar = ranges * np.cos(angles)
        y_lidar = ranges * np.sin(angles)
        z_lidar = np.zeros_like(x_lidar)

        # Stack into (N, 3) array
        points_lidar = np.vstack((x_lidar, y_lidar, z_lidar)).T

        return points_lidar, indices

    def transform_lidar_points_to_camera_frame(self, points_lidar, source_frame, target_frame, time_stamp):
        """
        Transform LiDAR points from source_frame to target_frame (e.g., camera frame).
        """
        try:
            # Use the timestamp from the scan message for accurate transformation
            transform = self.tf_buffer_.lookup_transform(
                target_frame=target_frame,
                source_frame=source_frame,
                time=time_stamp,
                timeout=rclpy.duration.Duration(seconds=1.0))
        except TransformException as ex:
            self.get_logger().warn(f'Could not transform {source_frame} to {target_frame}: {ex}')
            return None

        # Apply transformation
        translation = np.array([
            transform.transform.translation.x,
            transform.transform.translation.y,
            transform.transform.translation.z
        ])
        rotation = np.array([
            transform.transform.rotation.x,
            transform.transform.rotation.y,
            transform.transform.rotation.z,
            transform.transform.rotation.w
        ])
        rot_matrix = tf_transformations.quaternion_matrix(rotation)[:3, :3]
        points_camera = np.dot(rot_matrix, points_lidar.T).T + translation

        return points_camera
    
    # def transform_lidar_points_to_camera_frame(self, points_lidar, source_frame, target_frame, time_stamp):
    #     """
    #     Transform LiDAR points from source_frame to target_frame (e.g., camera frame).
    #     """
    #     try:
    #         # Use the timestamp from the scan message for accurate transformation
    #         transform = self.tf_buffer_.lookup_transform(
    #             target_frame=target_frame,
    #             source_frame=source_frame,
    #             time=time_stamp,
    #             timeout=rclpy.duration.Duration(seconds=1.0))
    #     except TransformException as ex:
    #         self.get_logger().warn(f'Could not transform {source_frame} to {target_frame}: {ex}')
    #         return None

    #     # Extract translation and rotation from the transform
    #     translation = np.array([
    #         transform.transform.translation.x,
    #         transform.transform.translation.y,
    #         transform.transform.translation.z
    #     ])
    #     rotation = np.array([
    #         transform.transform.rotation.x,
    #         transform.transform.rotation.y,
    #         transform.transform.rotation.z,
    #         transform.transform.rotation.w
    #     ])

    #     # Convert quaternion to rotation matrix using transforms3d
    #     rot_matrix = transforms3d.quaternions.quat2mat(rotation)

    #     # Apply transformation
    #     points_camera = np.dot(rot_matrix, points_lidar.T).T + translation

    #     return points_camera

    def project_points_to_image_plane(self, points_camera, fx, fy, cx, cy, image_msg, indices):
        """
        Project 3D points in camera frame onto the 2D image plane.
        """
        x_cam = points_camera[:, 0]
        y_cam = points_camera[:, 1]
        z_cam = points_camera[:, 2]

        # Avoid division by zero and points behind the camera
        valid_indices = z_cam > 0
        x_cam = x_cam[valid_indices]
        y_cam = y_cam[valid_indices]
        z_cam = z_cam[valid_indices]
        indices = indices[valid_indices]

        

        u = (fx * x_cam / z_cam + cx).astype(int)
        v = (fy * y_cam / z_cam + cy).astype(int)
        

        # Get image dimensions
        height, width = image_msg.height, image_msg.width

        # Filter points that are within the image frame
        valid_indices = (u >= 0) & (u < width) & (v >= 0) & (v < height)
        u = u[valid_indices]
        v = v[valid_indices]
        x_cam = x_cam[valid_indices]
        y_cam = y_cam[valid_indices]
        z_cam = z_cam[valid_indices]
        indices = indices[valid_indices]

        return u, v, x_cam, y_cam, z_cam, indices

    def associate_points_with_detections(self, u, v, x_cam, y_cam, z_cam, indices, labels, detections_msg, image_msg):
        """
        Associate projected points with YOLO detections and estimate object positions.
        """
        detections = detections_msg.detections
        pose_array_msg = PoseArray()
        pose_array_msg.header.stamp = self.get_clock().now().to_msg()
        pose_array_msg.header.frame_id = image_msg.header.frame_id  # Or robot base frame

        for idx, detection in enumerate(detections):
            class_id = detection.class_id

            # Process only red and green objects (assuming class IDs 0 and 1)
            if class_id in [1, 2]:
                # Get bounding box coordinates
                bbox = detection.bbox
                image_width = image_msg.width
                image_height = image_msg.height

                # Bounding box coordinates
                xmin = int(bbox.center.position.x - bbox.size.x / 2)
                ymin = int(bbox.center.position.y - bbox.size.y / 2)
                xmax = int(bbox.center.position.x + bbox.size.x / 2)
                ymax = int(bbox.center.position.y + bbox.size.y / 2)

                # Clip coordinates to image dimensions
                xmin = max(0, xmin)
                ymin = max(0, ymin)
                xmax = min(image_width - 1, xmax)
                ymax = min(image_height - 1, ymax)

                # Find points within bounding box
                in_bbox = (u >= xmin) & (u <= xmax) & (v >= ymin) & (v <= ymax)

                if np.any(in_bbox):
                    # Get corresponding 3D points in camera frame
                    x_obj = x_cam[in_bbox]
                    y_obj = y_cam[in_bbox]
                    z_obj = z_cam[in_bbox]

                    # Set labels for these points
                    labels[in_bbox] = class_id

                    # Estimate object position (e.g., mean of points)
                    x_mean = np.mean(x_obj)
                    y_mean = np.mean(y_obj)
                    z_mean = np.mean(z_obj)

                    # Create Pose message
                    pose_msg = Pose()
                    pose_msg.position.x = x_mean
                    pose_msg.position.y = y_mean
                    pose_msg.position.z = z_mean
                    # Orientation can be set if needed, else defaults to zero
                    pose_array_msg.poses.append(pose_msg)

                    # Log object position
                    # self.get_logger().info(
                    #     f"Detection {idx + 1}: Detected Object Position (x, y, z) = "
                    #     f"({x_mean:.2f}, {y_mean:.2f}, {z_mean:.2f}) meters"
                    # )
                else:
                    self.get_logger().warn("No LiDAR points found in bounding box for detection.")
            else:
                # Ignore detections that are not red or green
                continue

        return pose_array_msg, labels

    def caminfoCallback(self, msg: CameraInfo):
        """
        Callback function for handling camera information.
        """
        # Fill self.camera_info_ field
        K = np.array(msg.k)
        if len(K) == 9:
            K = K.reshape((3, 3))
            self.camera_info_ = {'fx': K[0][0], 'fy': K[1][1], 'cx': K[0][2], 'cy': K[1][2]}
            # self.get_logger().info("Camera info received and stored.")
            # self.get_logger().info(f"Camera Intrinsics - fx: {self.camera_info_['fx']}, fy: {self.camera_info_['fy']}, "
            #                        f"cx: {self.camera_info_['cx']}, cy: {self.camera_info_['cy']}")
        else:
            self.get_logger().warn("Invalid camera info received.")

def main(args=None):
    rclpy.init(args=args)
    node = Lidar2ImageNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: robot_perception/src/lidar2image_node.cpp
// lidar2image_node.cpp

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <yolov8_msgs/msg/detection_array.hpp>
#include <geometry_msgs/msg/pose_array.hpp>
#include <geometry_msgs/msg/pose.hpp>

#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <tf2_ros/message_filter.h>
#include <message_filters/subscriber.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <tf2/transform_datatypes.h>
// #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

#include <opencv2/opencv.hpp>

#include <vector>
#include <string>
#include <Eigen/Dense>
#include <numeric>
#include <algorithm>

class Lidar2ImageNode : public rclcpp::Node
{
public:
    Lidar2ImageNode();

private:
    // Subscribers
    message_filters::Subscriber<sensor_msgs::msg::Image> image_sub_;
    message_filters::Subscriber<yolov8_msgs::msg::DetectionArray> detections_sub_;
    message_filters::Subscriber<sensor_msgs::msg::LaserScan> scan_sub_;

    // Synchronizer
    typedef message_filters::sync_policies::ApproximateTime<
        yolov8_msgs::msg::DetectionArray,
        sensor_msgs::msg::Image,
        sensor_msgs::msg::LaserScan>
        SyncPolicy;
    message_filters::Synchronizer<SyncPolicy> sync_;

    // Camera info subscriber
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr caminfo_sub_;

    // Publishers
    rclcpp::Publisher<geometry_msgs::msg::PoseArray>::SharedPtr object_positions_pub_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr overlay_image_pub_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr labeled_scan_pub_;

    // TF buffer and listener
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

    // Camera intrinsics
    double fx_, fy_, cx_, cy_;
    bool camera_info_received_;

    // Callback functions
    void detection_image_scan_callback(
        const yolov8_msgs::msg::DetectionArray::ConstSharedPtr &detections_msg,
        const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
        const sensor_msgs::msg::LaserScan::ConstSharedPtr &scan_msg);

    void caminfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg);

    // Helper functions
    void compute_lidar_points(
        const sensor_msgs::msg::LaserScan::ConstSharedPtr &scan_msg,
        std::vector<Eigen::Vector3d> &points_lidar,
        std::vector<size_t> &indices);

    bool transform_lidar_points_to_camera_frame(
        const std::vector<Eigen::Vector3d> &points_lidar,
        const std::string &source_frame,
        const std::string &target_frame,
        const rclcpp::Time &time_stamp,
        std::vector<Eigen::Vector3d> &points_camera);

    void project_points_to_image_plane(
        const std::vector<Eigen::Vector3d> &points_camera,
        double fx, double fy, double cx, double cy,
        const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
        const std::vector<size_t> &indices,
        std::vector<int> &u, std::vector<int> &v,
        std::vector<double> &x_cam, std::vector<double> &y_cam, std::vector<double> &z_cam,
        std::vector<size_t> &valid_indices);

    void associate_points_with_detections(
        const std::vector<int> &u, const std::vector<int> &v,
        const std::vector<double> &x_cam, const std::vector<double> &y_cam, const std::vector<double> &z_cam,
        const std::vector<size_t> &valid_indices,
        std::vector<int> &labels,
        const yolov8_msgs::msg::DetectionArray::ConstSharedPtr &detections_msg,
        const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
        geometry_msgs::msg::PoseArray &pose_array_msg);
};

Lidar2ImageNode::Lidar2ImageNode()
    : Node("lidar2image_node"),
      image_sub_(this, "bgr_image"),
      detections_sub_(this, "/yolo/detections"),
      scan_sub_(this, "/scan"),
      sync_(SyncPolicy(100), detections_sub_, image_sub_, scan_sub_),
      camera_info_received_(false)
{
    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    sync_.registerCallback(std::bind(&Lidar2ImageNode::detection_image_scan_callback, this,
                                     std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));

    caminfo_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
        "camera_info", 10, std::bind(&Lidar2ImageNode::caminfoCallback, this, std::placeholders::_1));

    object_positions_pub_ = this->create_publisher<geometry_msgs::msg::PoseArray>("detected_object_positions", 10);
    overlay_image_pub_ = this->create_publisher<sensor_msgs::msg::Image>("overlay_image", 10);
    labeled_scan_pub_ = this->create_publisher<sensor_msgs::msg::LaserScan>("labeled_scan", 10);

    RCLCPP_INFO(this->get_logger(), "Initialized lidar2image_node");
}

void Lidar2ImageNode::detection_image_scan_callback(
    const yolov8_msgs::msg::DetectionArray::ConstSharedPtr &detections_msg,
    const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
    const sensor_msgs::msg::LaserScan::ConstSharedPtr &scan_msg)
{
    RCLCPP_INFO(this->get_logger(), "Processing synchronized messages");

    // Ensure camera_info_ is available
    if (!camera_info_received_)
    {
        RCLCPP_WARN(this->get_logger(), "Camera info not yet received. Cannot process data.");
        return;
    }

    // Get laser scan data and compute LiDAR points
    std::vector<Eigen::Vector3d> points_lidar;
    std::vector<size_t> indices;
    compute_lidar_points(scan_msg, points_lidar, indices);

    // Transform LiDAR points to camera frame
    rclcpp::Time time_stamp = scan_msg->header.stamp;
    std::vector<Eigen::Vector3d> points_camera;
    if (!transform_lidar_points_to_camera_frame(points_lidar, scan_msg->header.frame_id,
                                                image_msg->header.frame_id, time_stamp, points_camera))
    {
        RCLCPP_WARN(this->get_logger(), "Could not transform lidar points to camera");
        return;
    }

    // Project points onto image plane
    std::vector<int> u, v;
    std::vector<double> x_cam, y_cam, z_cam;
    std::vector<size_t> valid_indices;
    project_points_to_image_plane(points_camera, fx_, fy_, cx_, cy_, image_msg, indices,
                                  u, v, x_cam, y_cam, z_cam, valid_indices);

    // Initialize labels array
    std::vector<int> labels(u.size(), 0);

    // Associate projected points with detections and estimate object positions
    geometry_msgs::msg::PoseArray pose_array_msg;
    associate_points_with_detections(u, v, x_cam, y_cam, z_cam, valid_indices, labels,
                                     detections_msg, image_msg, pose_array_msg);

    // Publish the object positions
    object_positions_pub_->publish(pose_array_msg);

    // Convert image_msg to cv::Mat
    cv_bridge::CvImagePtr cv_ptr;
    try
    {
        cv_ptr = cv_bridge::toCvCopy(*image_msg, sensor_msgs::image_encodings::BGR8);
    }
    catch (cv_bridge::Exception &e)
    {
        RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
        return;
    }
    cv::Mat cv_image = cv_ptr->image;

    // Draw bounding boxes and labels for objects
    for (const auto &detection : detections_msg->detections)
    {
        int class_id = detection.class_id;

        if (class_id == 1 || class_id == 2 || class_id == 3)
        {
            auto bbox = detection.bbox;
            int image_width = image_msg->width;
            int image_height = image_msg->height;

            int xmin = static_cast<int>(bbox.center.position.x - bbox.size.x / 2);
            int ymin = static_cast<int>(bbox.center.position.y - bbox.size.y / 2);
            int xmax = static_cast<int>(bbox.center.position.x + bbox.size.x / 2);
            int ymax = static_cast<int>(bbox.center.position.y + bbox.size.y / 2);

            xmin = std::max(0, xmin);
            ymin = std::max(0, ymin);
            xmax = std::min(image_width - 1, xmax);
            ymax = std::min(image_height - 1, ymax);

            cv::Scalar color;
            std::string label_text;
            if (class_id == 1)
            {
                color = cv::Scalar(0, 255, 0); // Green
                label_text = "Green";
            }
            else if (class_id == 2)
            {
                color = cv::Scalar(203, 192, 255); // Pink
                label_text = "Parking";
            }
            else if (class_id == 3)
            {
                color = cv::Scalar(0, 0, 255); // Red
                label_text = "Red";
            }
            else
            {
                continue;
            }

            cv::rectangle(cv_image, cv::Point(xmin, ymin), cv::Point(xmax, ymax), color, 2);
            cv::putText(cv_image, label_text, cv::Point(xmin, ymin - 10),
                        cv::FONT_HERSHEY_SIMPLEX, 0.9, color, 2);
        }
    }

    // Draw labeled LiDAR points
    for (size_t i = 0; i < u.size(); ++i)
    {
        cv::Scalar color;
        if (labels[i] == 1)
        {
            color = cv::Scalar(0, 255, 0); // Green
        }
        else if (labels[i] == 2)
        {
            color = cv::Scalar(203, 192, 255); // Pink
        }
        else if (labels[i] == 3)
        {
            color = cv::Scalar(0, 0, 255); // Red
        }
        else
        {
            continue;
        }
        cv::circle(cv_image, cv::Point(u[i], v[i]), 2, color, -1);
    }

    // Publish the overlay image
    sensor_msgs::msg::Image::SharedPtr overlay_image_msg = cv_bridge::CvImage(
                                                              image_msg->header, sensor_msgs::image_encodings::BGR8, cv_image)
                                                              .toImageMsg();
    overlay_image_pub_->publish(*overlay_image_msg);

    // Create labels_for_scan array
    std::vector<float> labels_for_scan(scan_msg->ranges.size(), 0.0f);

    // Map labels back to the original scan indices
    for (size_t i = 0; i < valid_indices.size(); ++i)
    {
        size_t idx = indices[valid_indices[i]];
        labels_for_scan[idx] = static_cast<float>(labels[i]);
    }

    // Create new LaserScan message
    sensor_msgs::msg::LaserScan labeled_scan = *scan_msg;
    labeled_scan.intensities = labels_for_scan;

    // Publish the labeled scan
    labeled_scan_pub_->publish(labeled_scan);
}

void Lidar2ImageNode::compute_lidar_points(
    const sensor_msgs::msg::LaserScan::ConstSharedPtr &scan_msg,
    std::vector<Eigen::Vector3d> &points_lidar,
    std::vector<size_t> &indices)
{
    const auto &ranges = scan_msg->ranges;
    double angle = scan_msg->angle_min;

    for (size_t i = 0; i < ranges.size(); ++i)
    {
        double r = ranges[i];
        if (std::isfinite(r) && r >= scan_msg->range_min && r <= scan_msg->range_max)
        {
            double x = r * std::cos(angle);
            double y = r * std::sin(angle);
            double z = 0.0;
            points_lidar.emplace_back(x, y, z);
            indices.push_back(i);
        }
        angle += scan_msg->angle_increment;
    }
}

bool Lidar2ImageNode::transform_lidar_points_to_camera_frame(
    const std::vector<Eigen::Vector3d> &points_lidar,
    const std::string &source_frame,
    const std::string &target_frame,
    const rclcpp::Time &time_stamp,
    std::vector<Eigen::Vector3d> &points_camera)
{
    geometry_msgs::msg::TransformStamped transform;
    try
    {
        transform = tf_buffer_->lookupTransform(
            target_frame, source_frame, time_stamp, rclcpp::Duration::from_seconds(1.0));
    }
    catch (tf2::TransformException &ex)
    {
        RCLCPP_WARN(this->get_logger(), "Could not transform %s to %s: %s",
                    source_frame.c_str(), target_frame.c_str(), ex.what());
        return false;
    }

    Eigen::Translation3d translation(
        transform.transform.translation.x,
        transform.transform.translation.y,
        transform.transform.translation.z);

    Eigen::Quaterniond rotation(
        transform.transform.rotation.w,
        transform.transform.rotation.x,
        transform.transform.rotation.y,
        transform.transform.rotation.z);

    Eigen::Affine3d transform_matrix = translation * rotation;

    for (const auto &point : points_lidar)
    {
        Eigen::Vector3d point_transformed = transform_matrix * point;
        points_camera.push_back(point_transformed);
    }
    return true;
}

void Lidar2ImageNode::project_points_to_image_plane(
    const std::vector<Eigen::Vector3d> &points_camera,  // Use Eigen::Vector3d instead of Matrix<double, 3, 1>
    double fx, double fy, double cx, double cy,
    const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
    const std::vector<size_t> &indices,
    std::vector<int> &u, std::vector<int> &v,
    std::vector<double> &x_cam, std::vector<double> &y_cam, std::vector<double> &z_cam,
    std::vector<size_t> &valid_indices)
{
    int width = image_msg->width;
    int height = image_msg->height;

    for (const auto& point : points_camera) {
        double x = point.x();  // Access the x component
        double y = point.y();  // Access the y component
        double z = point.z();  // Access the z component

        if (z > 0) {  // Only consider points in front of the camera
            int ui = static_cast<int>(fx * x / z + cx);
            int vi = static_cast<int>(fy * y / z + cy);

            if (ui >= 0 && ui < width && vi >= 0 && vi < height) {
                u.push_back(ui);
                v.push_back(vi);
                x_cam.push_back(x);
                y_cam.push_back(y);
                z_cam.push_back(z);
                valid_indices.push_back(&point - &points_camera[0]);  // Index of the point
            }
        }
    }
}


void Lidar2ImageNode::associate_points_with_detections(
    const std::vector<int> &u, const std::vector<int> &v,
    const std::vector<double> &x_cam, const std::vector<double> &y_cam, const std::vector<double> &z_cam,
    const std::vector<size_t> &valid_indices,
    std::vector<int> &labels,
    const yolov8_msgs::msg::DetectionArray::ConstSharedPtr &detections_msg,
    const sensor_msgs::msg::Image::ConstSharedPtr &image_msg,
    geometry_msgs::msg::PoseArray &pose_array_msg)
{
    pose_array_msg.header.stamp = this->now();
    pose_array_msg.header.frame_id = image_msg->header.frame_id;

    for (const auto &detection : detections_msg->detections)
    {
        int class_id = detection.class_id;

        if (class_id == 1 || class_id == 2 || class_id == 3)
        {
            auto bbox = detection.bbox;
            int image_width = image_msg->width;
            int image_height = image_msg->height;

            int xmin = static_cast<int>(bbox.center.position.x - bbox.size.x / 2);
            int ymin = static_cast<int>(bbox.center.position.y - bbox.size.y / 2);
            int xmax = static_cast<int>(bbox.center.position.x + bbox.size.x / 2);
            int ymax = static_cast<int>(bbox.center.position.y + bbox.size.y / 2);

            xmin = std::max(0, xmin);
            ymin = std::max(0, ymin);
            xmax = std::min(image_width - 1, xmax);
            ymax = std::min(image_height - 1, ymax);

            // Find points within bounding box
            std::vector<double> x_obj, y_obj, z_obj;

            for (size_t i = 0; i < u.size(); ++i)
            {
                if (u[i] >= xmin && u[i] <= xmax && v[i] >= ymin && v[i] <= ymax)
                {
                    x_obj.push_back(x_cam[i]);
                    y_obj.push_back(y_cam[i]);
                    z_obj.push_back(z_cam[i]);
                    labels[i] = class_id;
                }
            }

            if (!x_obj.empty())
            {
                // Estimate object position
                double x_mean = std::accumulate(x_obj.begin(), x_obj.end(), 0.0) / x_obj.size();
                double y_mean = std::accumulate(y_obj.begin(), y_obj.end(), 0.0) / y_obj.size();
                double z_mean = std::accumulate(z_obj.begin(), z_obj.end(), 0.0) / z_obj.size();

                geometry_msgs::msg::Pose pose_msg;
                pose_msg.position.x = x_mean;
                pose_msg.position.y = y_mean;
                pose_msg.position.z = z_mean;
                pose_array_msg.poses.push_back(pose_msg);
            }
            else
            {
                RCLCPP_WARN(this->get_logger(), "No LiDAR points found in bounding box for detection.");
            }
        }
    }
}

void Lidar2ImageNode::caminfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
{
    const auto &K = msg->k;
    if (K.size() == 9)
    {
        fx_ = K[0];
        fy_ = K[4];
        cx_ = K[2];
        cy_ = K[5];
        camera_info_received_ = true;

        RCLCPP_INFO(this->get_logger(), "Camera info received and stored.");
    }
    else
    {
        RCLCPP_WARN(this->get_logger(), "Invalid camera info received.");
    }
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<Lidar2ImageNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
